### PR TITLE
Change lifetime of Socket used in NamedPipeServerStream on Unix

### PR DESCRIPTION
**Pull request by stephentoub**
We currently create and destroy the unix domain socket each time WaitForConnection{Async} is called, and it lives until the client connects.  This causes problems when multiple clients try to connect before WaitForConnection{Async} is called; any number that's under the min listen limit applied by the OS will have their connections completed, even though Accept hasn't been called, and then when the socket is closed, the connections will all fail.  This change moves the UDS creation to when the NamedPipeServerStream is created, deleting it when the NPSS is disposed.

cc: @pjanotti, @JeremyKuhne, @jaredpar, @khyperia

### CODE REVIEW

**Comment by stephentoub** [comment]
Hold off on reviewing this.  Two things:
1. macOS is having some issues with the change and I need to investigate.
2. Now that the Socket is around for the duration of the stream, we may be able to support concurrent NamedPipeServerStream instances, as Window does.

**Comment by stephentoub** [comment]
@dotnet/dnceng, my OSX leg has:
https://ci3.dot.net/job/dotnet_corefx/job/master/job/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/4558/consoleText
```
All nodes of label ‘osx-10.12||OSX.1012.Amd64.Open’ are offline
```
Is this expected?

**Comment by mmitche** [comment]
@stephentoub We're low on nodes right now because of cleanup issues.  @ravimeda is working to get them back online.

**Comment by stephentoub** [comment]
Thanks, @mmitche.

**Comment by stephentoub** [comment]
Ok, this can be reviewed now.  I added a commit that implements maxNumberOfServerInstances on Unix.  It's still a subset of what Windows provides, in that Windows supports this across processes, but at least now a Unix process can create multiple server streams that all coordinate with each other and respect maxNumberOfServerInstances, such that a server process can handle multiple client requests concurrently.

File: src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.Specific.cs [hunk 1]
```diff
@@ -77,6 +79,147 @@ public void ConnectWithConflictingDirections_Throws_UnauthorizedAccessException(
...
+                    await Task.WhenAll(writeAsync, readAsync);
+                    Assert.Equal(1, await readAsync);
+
+                    server.Disconnect();
+                }
+            }
+        }
+
+        [Fact]
+        public void MaxNumberOfServerInstances_TooManyServers_Throws()
```
Comment (pjanotti) [inline src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.Specific.cs]: This test only made sense to me in light of the comments at `SharedServer.Get` perhaps a brief mention here can be helpful.

Comment (stephentoub) [reply src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.Specific.cs]: Ok, will add some comments.

**Comment by khyperia** [comment]
I just discovered another behavior of NamedPipeServerStream that we are depending on in Roslyn:

When `WaitForConnection()` is being called, and (on another stream) NamedPipeServerStream.Close() is called: on Windows, the WaitForConnection fails immediately, and on Linux, it hangs infinitely (as the server `Socket` was never closed, it's a local variable). More generally though, it is impossible (as far as me and Jared can tell) to cancel an in-progress `WaitForConnection` (cancellationtokens aren't used, etc.)

It would be nice if you implemented a way to cancel a `WaitForConnection` (hopefully by just implementing the behavior that `Close` cancels it).

(Note I haven't looked through this PR to see if it does so, so my apologies if it already does. Also, let me know if I should report an actual issue for this.)

**Comment by stephentoub** [comment]
> I just discovered another behavior of NamedPipeServerStream that we are depending on in Roslyn

Can you please open a separate issue with a small repro?

File: src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.Specific.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.Specific.cs]: Added

Comment (pjanotti) [reply src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.Specific.cs]: Thanks.

**Comment by stephentoub** [comment]
@dotnet-bot test  please (https://github.com/dotnet/corefx/issues/23341, https://github.com/dotnet/corefx/issues/24874)

@dotnet-bot test OSX x64 Debug Build please (trying to get it to fail)
@dotnet-bot test Linux x64 Debug Build please (trying to get it to fail)
@dotnet-bot test Linux x64 Release Build please (trying to get it to fail)

**Comment by stephentoub** [comment]
@dotnet-bot test Windows x64 Debug Build please (#23341, #24874)

**Comment by stephentoub** [comment]
@dotnet-bot test OSX x64 Debug Build please (trying to get it to fail)
@dotnet-bot test Linux x64 Debug Build please (trying to get it to fail)
@dotnet-bot test Linux x64 Release Build please (trying to get it to fail)

**Comment by stephentoub** [comment]
@dotnet-bot test OSX x64 Debug Build please (trying to get it to fail)
@dotnet-bot test Linux x64 Debug Build please (trying to get it to fail)
@dotnet-bot test Linux x64 Release Build please (trying to get it to fail)

**Pull request merged**